### PR TITLE
[App] Refactor cloud dispatch and update to new API

### LINKED ISSUE #15012: Split out `CloudRuntime.dispatch` into multiple methods
## Proposed refactor

Split out CloudRuntime.dispatch into multiple methods. 

### Motivation

This makes the code easier to read and test. Specifically, when testing individual setup stages of this runtime, we can mock out individual parts that are irrelevant for execution/testing.

### Additional context

This first came up here: #14929

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging PyTorch Lightning, Transformers, and Hydra.

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

- Updates cloud dispatch to the new API
- breaks down the large method into several smaller ones and calls them in a consistent order (e.g. so we don't raise errors after things have started to happen)
- adds the option to export `LIGHTNING_CLOUD_PRINT_SPECS=gallery` or `LIGHTNING_CLOUD_PRINT_SPECS=web` to generate app specs for use in the gallery or web dispatch

Note: Tests left deliberately as unchanged as possible to ensure that the core logic hasn't changed.

Fixes #15012

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: requirements/app/base.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-lightning-cloud>=0.5.12, <=0.5.16
+lightning-cloud==0.5.19
```
Comment (carmocca) [inline requirements/app/base.txt:1]:
```suggestion
lightning-cloud>=0.5.19
```
This will avoid future headaches with upgrades

**Review by carmocca — APPROVED** [review]

File: src/lightning_app/runners/cloud.py [hunk 2]
```diff
@@ -243,367 +137,622 @@ def dispatch(
...
+            network_configs = self._get_network_configs(flow_servers)
+            works = self._get_works()
+            run_body = self._get_run_body(
+                cluster_id, flow_servers, network_configs, works, no_cache, root, self.start_server
+            )
+            auth = self._get_auth(self.enable_basic_auth)
+            env_vars = self._get_env_vars(self.env_vars, self.secrets, self.run_app_comment_commands)
 
-        return random.choice(clusters).id
+            if LIGHTNING_CLOUD_PRINT_SPECS:
```
Comment (awaelchli) [inline src/lightning_app/runners/cloud.py:695]:
```suggestion
            if LIGHTNING_CLOUD_PRINT_SPECS is not None:
```

File: src/lightning_app/source_code/local.py [hunk 3]
```diff
@@ -27,6 +28,14 @@ def __init__(self, path: Path, ignore_functions: Optional[List[_IGNORE_FUNCTION]
         if not self.cache_location.exists():
             self.cache_location.mkdir(parents=True, exist_ok=True)
 
+        # Create a default dotignore if it doesn't exist
```
Comment (awaelchli) [inline src/lightning_app/source_code/local.py:31]: Let's have a test for this new behavior?

File: src/lightning_app/runners/cloud.py [hunk 4]
```diff
@@ -243,367 +137,622 @@ def dispatch(
...
+        logger.info(f"flow_servers: {flow_servers_json}")
+        works_json = json.dumps(_to_clean_dict(run_body.works, print_format == "web"), separators=(",", ":"))
+        logger.info(f"works: {works_json}")
+        logger.info(f"entrypoint_file: {run_body.app_entrypoint_file}")
+        requirements_path = getattr(getattr(run_body.image_spec, "dependency_file_info", ""), "path", "")
+        logger.info(f"requirements_path: {requirements_path}")
 
-        if len(cluster_bindings) == 1:
-            return cluster_bindings[0].cluster_id
+    def dispatch(
```
Comment (awaelchli) [inline src/lightning_app/runners/cloud.py:636]: I suggest putting the public methods at the top of the class

**Review by awaelchli — APPROVED** [review]
I think this addresses #15012 too. Would you agree?

File: src/lightning_app/source_code/local.py (hunk 3 above)
Comment (ethanwharris) [reply src/lightning_app/source_code/local.py:31]: Was added in #16056 along with a test, just moving it in this PR 😃

File: src/lightning_app/runners/cloud.py (hunk 4 above)
Comment (ethanwharris) [reply src/lightning_app/runners/cloud.py]: Done 😃
